adam/visualization/positioning.py

```diff
@@ -0,0 +1,426 @@
+    e.g. a box centered at (0, 0, 0) and with a scale of (1, 1, 1) would have opposite
+    corners at (-1, -1, -1) and (1, 1, 1), giving the box a volume of 2(^3)
+    """
+
+    center: torch.Tensor = attrib()  # tensor shape: (3,)
+    scale: torch.Tensor = attrib()  # tensor shape: (3, 3) - diagonal matrix
+    # rotation: torch.Tensor = attrib()
+
+    def center_distance_from_point(self, point: torch.Tensor) -> torch.Tensor:
+        return torch.dist(self.center, point, 2)
+
+    @staticmethod
+    def create_at_random_position(
+        *, min_distance_from_origin: float, max_distance_from_origin: float
+    ):
+        return AxisAlignedBoundingBox.create_at_random_position_scaled(
+            min_distance_from_origin=min_distance_from_origin,
+            max_distance_from_origin=max_distance_from_origin,
+            object_scale=torch.ones(3),
+        )
+
+    @staticmethod
+    def create_at_center_point(*, center: ndarray):
+        return AxisAlignedBoundingBox(
+            Parameter(
+                torch.tensor(center, dtype=torch.float),  # pylint: disable=not-callable
+                requires_grad=True,
+            ),
+            torch.diag(torch.ones(3)),
+        )
+
+    @staticmethod
+    def create_at_random_position_scaled(
+        *,
+        min_distance_from_origin: float,
+        max_distance_from_origin: float,
+        object_scale: torch.Tensor,
+    ):
+        check_arg(min_distance_from_origin > 0.0)
+        check_arg(min_distance_from_origin < max_distance_from_origin)
+        # we first generate a random point on the unit sphere by
+        # generating a random vector in cube...
+        center = np.random.randn(3, 1).squeeze()
+        # and then normalizing.
+        center /= np.linalg.norm(center)
+
+        # then we scale according to the distances above
+        scale_factor = np.random.uniform(
+            min_distance_from_origin, max_distance_from_origin
+        )
+        center *= scale_factor
+        return AxisAlignedBoundingBox(
+            Parameter(
+                torch.tensor(center, dtype=torch.float),  # pylint: disable=not-callable
+                requires_grad=True,
+            ),
+            torch.diag(object_scale),
+        )
+
+    def get_corners(self) -> torch.Tensor:
+        return self.center.expand(8, 3) + torch.tensor(  # pylint: disable=not-callable
+            [
+                [-1, -1, -1],
+                [1, -1, -1],
+                [-1, 1, -1],
+                [1, 1, -1],
+                [-1, -1, 1],
+                [1, -1, 1],
+                [-1, 1, 1],
+                [1, 1, 1],
+            ],
+            dtype=torch.float,
+        ).matmul(self.scale)
+        # see https://github.com/pytorch/pytorch/issues/24807 re: pylint issue
+
+    # helper functions giving names to a few corners used in calculations:
+    def right_corner(self) -> torch.Tensor:
```

Comment:
These three `X_corner` functions are used in only one place each; it might be better to just in-line them there unless you expect a general need for them.